docs/proposals/config_data.md

```diff
@@ -0,0 +1,315 @@
+is:
+
+1. A `ConfigData` object is created
+2. A pod that consumes the configuration data via environment variables is created
+3. The pod is scheduled onto a node
+4. The kubelet retrieves the `ConfigData` resource(s) referenced by the pod and starts the container
+   processes with the appropriate data in environment variables
+
+### Consuming `ConfigData` in Volumes
+
+Many programs read their configuration from configuration files.  `ConfigData` should be possible
+to consume in a volume.  The rough series of events for consuming `ConfigData` this way
+is:
+
+1. A `ConfigData` object is created
+2. A new pod using the `ConfigData` via the volume plugin is created
+3. The pod is scheduled onto a node
+4. The Kubelet creates an instance of the volume plugin and calls its `Setup()` method
+5. The volume plugin retrieves the `ConfigData` resource(s) referenced by the pod and projects
+   the appropriate data into the volume
+
+### Consuming `ConfigData`  Updates
+
+Any long-running system has configuration that is mutated over time.  Changes made to configuration
+data must be made visible to pods consuming data in volumes so that they can respond to those
+changes.
+
+The `resourceVersion` of the `ConfigData` object will be updated by the API server every time the
+object is modified.  After an update, modifications will be made visible to the consumer container:
+
+1. A `ConfigData` object is created
+2. A new pod using the `ConfigData` via the volume plugin is created
+3. The pod is scheduled onto a node
+4. During the sync loop, the Kubelet creates an instance of the volume plugin and calls its
+   `Setup()` method
+5. The volume plugin retrieves the `ConfigData` resource(s) referenced by the pod and projects
+   the appropriate data into the volume
+6. The `ConfigData` referenced by the pod is updated
+7. During the next iteration of the `syncLoop`, the Kubelet creates an instance of the volume plugin
+   and calls its `Setup()` method
+8. The volume plugin projects the updated data into the volume
+
+It is the consuming pod's responsibility to make use of the updated data once it is made visible.
+
+Because environment variables cannot be updated without restarting a container, configuration data
+consumed in environment variables will not be updated.
+
+### Advantages
+
+* Easy to consume in pods; consumer-agnostic
+* Configuration data is persistent and versioned
+* Consumers of configuration data in volumes can respond to changes in the data
+
+## Proposed Design
+
+### API Resource
+
+The `ConfigData` resource will be added to the `extensions` API Group:
+
+```go
+package api
+
+// ConfigData holds configuration data for pods to consume.
+type ConfigData struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+  // Data contains the configuration data.  Each key must be a valid DNS_SUBDOMAIN or leading
+  // dot followed by valid DNS_SUBDOMAIN.
+	Data map[string]string `json:"data,omitempty"`
```

Comment:
@thockin at our earlier discussions (now disappeared in github) we've decided that map[string]string is sufficient for server our use cases. AFAIK there aren't any common config format that is binary in nature apart from encrypted files and certificates (also available in text format). For most configuration files seem like text format is enough. Granted it makes binary data a bit tricky, is this a common concern that we should limit the ConfigData for?

For your "dot followed by DNS_SUBDOMAIN" comment, I agree and we can do something like a valid DNS_SUBDOMAIN for key names and then use the desired file names  as ``path: /etc/.gitconfig``, ``path: /etc/redis.conf`` etc.

For the ``crossover`` part, I think such logic would be bloating ``ConfigData``. Instead, that part should perhaps be discussed for advanced rolling update logic or in a similar context.